File: Assemblies/EonicWebV5/Tools/Integration/Integration.Directory.Twitter.vb
Option Strict Off
Option Explicit On

Imports System.Collections.Specialized
Imports Eonic.Tools.Http.Utils
Imports System


Namespace Integration.Directory

    ''' <summary>
    ''' Twitter class
    ''' </summary>
    ''' <remarks></remarks>
    Public Class Twitter

        Inherits Integration.Directory.BaseProvider
        Implements Integration.Directory.IPostable

        Public twitterConsumerKey As String = "c2h0VNmcE5c0Vu0fsHEfGg"
        Public twitterConsumerSecret As String = "5DzFnQtX7cNxuChpJCcz1ZzhGfPGTAh4MYWMuuJvxQ"

        Private Const _postLengthLimit As Integer = 140

        Public Shadows Event OnError(ByVal sender As Object, ByVal e As Eonic.Tools.Errors.ErrorEventArgs)


        Public Sub New(ByRef aWeb As Eonic.Web)
            MyBase.New(aWeb)
        End Sub
        Public Sub New(ByRef aWeb As Eonic.Web, ByRef directoryId As Long)
            MyBase.New(aWeb, directoryId)
        End Sub

        Public Function GetRequestToken() As String

            Try
                If IsAuthorisedUser Then


                    Dim callbackParameters As New NameValueCollection(3)
                    callbackParameters.Add("oAuthResp", "twitter")
                    callbackParameters.Add("integration", "Twitter.AccessTokens")
                    callbackParameters.Add("dirId", MyBase.DirectoryId.ToString())

                    Dim callback As Uri = BuildURIFromRequest(myWeb.moRequest, callbackParameters)

                    Dim twitterAPI As New Eonic.Tools.Integration.Twitter.TwitterVB2.TwitterAPI()
                    Dim authenticationLink As String = twitterAPI.GetAuthenticationLink(twitterConsumerKey, twitterConsumerSecret, callback.ToString())

                    myWeb.AddResponse("Twitter.AuthenticationLink", authenticationLink, , ResponseType.Redirect)
                    myWeb.msRedirectOnEnd = authenticationLink
                    Return authenticationLink
                Else

                    myWeb.AddResponse("Twitter.GetRequestToken.Unauthorised", "The user requesting this method is not authorised", , ResponseType.Alert)
                    Return ""

                End If

            Catch ex As Exception
                RaiseEvent OnError(Me, New Eonic.Tools.Errors.ErrorEventArgs(_moduleName, "GetRequestToken", ex, ""))
                Return ""
            End Try


        End Function

        Public Sub AccessTokens()
            Try
                Dim token As String = myWeb.moRequest("oauth_token")
                Dim verifier As String = myWeb.moRequest("oauth_verifier")
                AccessTokens(token, verifier)
            Catch ex As Exception
                RaiseEvent OnError(Me, New Eonic.Tools.Errors.ErrorEventArgs(_moduleName, "AccessTokens()", ex, ""))
            End Try
        End Sub
        Public Sub AccessTokens(ByVal token As String, ByVal verifier As String)

            Try

                If IsAuthorisedUser Then
                    Dim twitterAPI As New Eonic.Tools.Integration.Twitter.TwitterVB2.TwitterAPI()

                    If String.IsNullOrEmpty(token) Or String.IsNullOrEmpty(verifier) Then
                        ' token or Verifier are not populated
                        myWeb.AddResponse("Twitter.AccessTokens.NullArguments", "Either the token or the verifier were blank", , ResponseType.Alert)

                    Else
                        twitterAPI.GetAccessTokens(twitterConsumerKey, twitterConsumerSecret, token, verifier)

                        _credentials.AddSetting("OAuth_Token", twitterAPI.OAuth_Token)
                        _credentials.AddSetting("OAuth_TokenSecret", twitterAPI.OAuth_TokenSecret)

                        ' Get the user name
                        Try
                            Dim user As Eonic.Tools.Integration.Twitter.TwitterVB2.TwitterUser
                            twitterAPI.AuthenticateWith(twitterConsumerKey, twitterConsumerSecret, twitterAPI.OAuth_Token, twitterAPI.OAuth_TokenSecret)
                            user = twitterAPI.AccountInformation()
                            _credentials.AddSetting("Name", user.Name)
                            _credentials.AddSetting("ScreenName", user.ScreenName)
                        Catch ex As Exception

                        End Try

                        SaveCredentials()

                        myWeb.AddResponse("Twitter.AccessTokens.Success", "User account has been successfully linked to Twitter", , ResponseType.Hint)

                    End If
                Else

                    myWeb.AddResponse("Twitter.AccessTokens.Unauthorised", "The user requesting this method is not authorised", , ResponseType.Alert)

                End If



            Catch ex As Exception
                RaiseEvent OnError(Me, New Eonic.Tools.Errors.ErrorEventArgs(_moduleName, "AccessTokens(String,String)", ex, ""))
            End Try

        End Sub

        Public Sub Update()
            Try


                Dim status As String = myWeb.moRequest("status")
                Update(status)

            Catch ex As Exception
                RaiseEvent OnError(Me, New Eonic.Tools.Errors.ErrorEventArgs(_moduleName, "Update(String)", ex, ""))
            End Try

        End Sub
        Public Sub Update(ByVal status As String)
            Try


                If Not IsAuthorisedUser Then
                    myWeb.AddResponse("Twitter.Update.Unauthorised", "The user requesting this method is not authorised", , ResponseType.Alert)

                ElseIf Not Me.IsProviderCredentialLoaded Then
                    myWeb.AddResponse("Twitter.Update.NoCredentials", "Could not load credentials for user Id" & DirectoryId, , ResponseType.Alert)

                Else
                    ' Load the credentials
                    Dim token As String = _credentials.GetSetting("OAuth_Token")
                    Dim secret As String = _credentials.GetSetting("OAuth_TokenSecret")

                    If String.IsNullOrEmpty(token) Or String.IsNullOrEmpty(token) Then
                        myWeb.AddResponse("Twitter.Update.NullArguments", "No valid credentials are available", , ResponseType.Alert)

                    ElseIf String.IsNullOrEmpty(status) Then
                        myWeb.AddResponse("Twitter.Update.NullStatus", "The status is blank", , ResponseType.Alert)

                    Else
                        Dim twitterAPI As New Eonic.Tools.Integration.Twitter.TwitterVB2.TwitterAPI()
                        twitterAPI.AuthenticateWith(twitterConsumerKey, twitterConsumerSecret, token, secret)
                        twitterAPI.Update(status)
                    End If


                End If

            Catch ex As Exception
                RaiseEvent OnError(Me, New Eonic.Tools.Errors.ErrorEventArgs(_moduleName, "Update(String)", ex, ""))
                myWeb.AddResponse("Twitter.Update.Failed", "The update failed - see error log for more detail", , ResponseType.Alert)
            End Try
        End Sub



        Public Overrides Sub Post(ByVal content As String, Optional ByVal trackbackUri As Uri = Nothing, Optional ByVal contentId As Long = 0)
            Try

                ' Is there something to post
                If Not String.IsNullOrEmpty(content) Then

                    Dim postLimitLength As Integer = _postLengthLimit

                    ' First shorten the URL, if it exists.
                    Dim shortenedURL As String = ""
                    If trackbackUri IsNot Nothing Then
                        shortenedURL = Tools.Http.Utils.ShortenURL(trackbackUri.ToString)
                        ' If shortening has failed check the original uri
                        If String.IsNullOrEmpty(shortenedURL) Then shortenedURL = trackbackUri.ToString
                        If Not String.IsNullOrEmpty(shortenedURL) Then shortenedURL = " " & shortenedURL
                    End If

                    postLimitLength -= shortenedURL.Length

                    ' Truncate the content
                    content = Tools.Text.TruncateString(content, postLimitLength)

                    Me.Update(content & shortenedURL)

                    myWeb.moDbHelper.logActivity(Web.dbHelper.ActivityType.IntegrationTwitterPost, Me.DirectoryId, 0, contentId, shortenedURL)

                End If

            Catch ex As Exception
                RaiseEvent OnError(Me, New Eonic.Tools.Errors.ErrorEventArgs(_moduleName, "Post()", ex, ""))
            End Try
        End Sub

        Public Class User
            Public Property id() As String
                Get
                    Return m_id
                End Get
                Set
                    m_id = Value
                End Set
            End Property
            Private m_id As String
            Public Property first_name() As String
                Get
                    Return m_first_name
                End Get
                Set
                    m_first_name = Value
                End Set
            End Property
            Private m_first_name As String
            Public Property last_name() As String
                Get
                    Return m_last_name
                End Get
                Set
                    m_last_name = Value
                End Set
            End Property
            Private m_last_name As String
            Public Property email() As String
                Get
                    Return m_email
                End Get
                Set
                    m_email = Value
                End Set
            End Property
            Private m_email As String
            Public Property link() As String
                Get
                    Return m_link
                End Get
                Set
                    m_link = Value
                End Set
            End Property
            Private m_link As String
            Public Property username() As String
                Get
                    Return m_username
                End Get
                Set
                    m_username = Value
                End Set
            End Property
            Private m_username As String
            Public Property gender() As String
                Get
                    Return m_gender
                End Get
                Set
                    m_gender = Value
                End Set
            End Property
            Private m_gender As String
            Public Property locale() As String
                Get
                    Return m_locale
                End Get
                Set
                    m_locale = Value
                End Set
            End Property
            Private m_locale As String
        End Class

    End Class



End Namespace

File: Assemblies/ProteanCMS/Tools/Integration/Integration.Directory.Twitter.cs
using System;
using System.Collections.Specialized;
using static Protean.stdTools;
using static Protean.Tools.Http.Utils;


namespace Protean.Integration.Directory
{

    /// <summary>
    /// Twitter class
    /// </summary>
    /// <remarks></remarks>

    public class Twitter : Directory.BaseProvider, Directory.IPostable
    {

        public string twitterConsumerKey = "c2h0VNmcE5c0Vu0fsHEfGg";
        public string twitterConsumerSecret = "5DzFnQtX7cNxuChpJCcz1ZzhGfPGTAh4MYWMuuJvxQ";

        private const int _postLengthLimit = 140;

        public new event OnErrorEventHandler OnError;

        public new delegate void OnErrorEventHandler(object sender, Tools.Errors.ErrorEventArgs e);


        public Twitter(ref Cms aWeb) : base(ref aWeb)
        {
        }
        public Twitter(ref Cms aWeb, ref long directoryId) : base(ref aWeb, ref directoryId)
        {
        }

        public string GetRequestToken()
        {

            try
            {
                if (this.IsAuthorisedUser)
                {


                    var callbackParameters = new NameValueCollection(3);
                    callbackParameters.Add("oAuthResp", "twitter");
                    callbackParameters.Add("integration", "Twitter.AccessTokens");
                    callbackParameters.Add("dirId", base.DirectoryId.ToString());

                    var callback = Tools.Http.Utils.BuildURIFromRequest(this.myWeb.moRequest, callbackParameters);

                    var twitterAPI = new Tools.Integration.Twitter.TwitterVB2.TwitterAPI();
                    string authenticationLink = twitterAPI.GetAuthenticationLink(twitterConsumerKey, twitterConsumerSecret, callback.ToString());

                    this.myWeb.AddResponse("Twitter.AuthenticationLink", authenticationLink, default, ResponseType.Redirect);
                    this.myWeb.msRedirectOnEnd = authenticationLink;
                    return authenticationLink;
                }
                else
                {

                    this.myWeb.AddResponse("Twitter.GetRequestToken.Unauthorised", "The user requesting this method is not authorised", default, ResponseType.Alert);
                    return "";

                }
            }

            catch (Exception ex)
            {
                OnError?.Invoke(this, new Tools.Errors.ErrorEventArgs(this._moduleName, "GetRequestToken", ex, ""));
                return "";
            }


        }

        public void AccessTokens()
        {
            try
            {
                string token = this.myWeb.moRequest("oauth_token");
                string verifier = this.myWeb.moRequest("oauth_verifier");
                AccessTokens(token, verifier);
            }
            catch (Exception ex)
            {
                OnError?.Invoke(this, new Tools.Errors.ErrorEventArgs(this._moduleName, "AccessTokens()", ex, ""));
            }
        }
        public void AccessTokens(string token, string verifier)
        {

            try
            {

                if (this.IsAuthorisedUser)
                {
                    var twitterAPI = new Tools.Integration.Twitter.TwitterVB2.TwitterAPI();

                    if (string.IsNullOrEmpty(token) | string.IsNullOrEmpty(verifier))
                    {
                        // token or Verifier are not populated
                        this.myWeb.AddResponse("Twitter.AccessTokens.NullArguments", "Either the token or the verifier were blank", default, ResponseType.Alert);
                    }

                    else
                    {
                        twitterAPI.GetAccessTokens(twitterConsumerKey, twitterConsumerSecret, token, verifier);

                        this._credentials.AddSetting("OAuth_Token", twitterAPI.OAuth_Token);
                        this._credentials.AddSetting("OAuth_TokenSecret", twitterAPI.OAuth_TokenSecret);

                        // Get the user name
                        try
                        {
                            Tools.Integration.Twitter.TwitterVB2.TwitterUser user;
                            twitterAPI.AuthenticateWith(twitterConsumerKey, twitterConsumerSecret, twitterAPI.OAuth_Token, twitterAPI.OAuth_TokenSecret);
                            user = twitterAPI.AccountInformation();
                            this._credentials.AddSetting("Name", user.Name);
                            this._credentials.AddSetting("ScreenName", user.ScreenName);
                        }
                        catch (Exception ex)
                        {

                        }

                        this.SaveCredentials();

                        this.myWeb.AddResponse("Twitter.AccessTokens.Success", "User account has been successfully linked to Twitter", default, ResponseType.Hint);

                    }
                }
                else
                {

                    this.myWeb.AddResponse("Twitter.AccessTokens.Unauthorised", "The user requesting this method is not authorised", default, ResponseType.Alert);

                }
            }



            catch (Exception ex)
            {
                OnError?.Invoke(this, new Tools.Errors.ErrorEventArgs(this._moduleName, "AccessTokens(String,String)", ex, ""));
            }

        }

        public void Update()
        {
            try
            {


                string status = this.myWeb.moRequest("status");
                Update(status);
            }

            catch (Exception ex)
            {
                OnError?.Invoke(this, new Tools.Errors.ErrorEventArgs(this._moduleName, "Update(String)", ex, ""));
            }

        }
        public void Update(string status)
        {
            try
            {


                if (!this.IsAuthorisedUser)
                {
                    this.myWeb.AddResponse("Twitter.Update.Unauthorised", "The user requesting this method is not authorised", default, ResponseType.Alert);
                }

                else if (!this.IsProviderCredentialLoaded)
                {
                    this.myWeb.AddResponse("Twitter.Update.NoCredentials", "Could not load credentials for user Id" + this.DirectoryId, default, ResponseType.Alert);
                }

                else
                {
                    // Load the credentials
                    string token = this._credentials.GetSetting("OAuth_Token");
                    string secret = this._credentials.GetSetting("OAuth_TokenSecret");

                    if (string.IsNullOrEmpty(token) | string.IsNullOrEmpty(token))
                    {
                        this.myWeb.AddResponse("Twitter.Update.NullArguments", "No valid credentials are available", default, ResponseType.Alert);
                    }

                    else if (string.IsNullOrEmpty(status))
                    {
                        this.myWeb.AddResponse("Twitter.Update.NullStatus", "The status is blank", default, ResponseType.Alert);
                    }

                    else
                    {
                        var twitterAPI = new Tools.Integration.Twitter.TwitterVB2.TwitterAPI();
                        twitterAPI.AuthenticateWith(twitterConsumerKey, twitterConsumerSecret, token, secret);
                        twitterAPI.Update(status);
                    }


                }
            }

            catch (Exception ex)
            {
                OnError?.Invoke(this, new Tools.Errors.ErrorEventArgs(this._moduleName, "Update(String)", ex, ""));
                this.myWeb.AddResponse("Twitter.Update.Failed", "The update failed - see error log for more detail", default, ResponseType.Alert);
            }
        }



        public override void Post(string content, Uri trackbackUri = null, long contentId = 0L)
        {
            try
            {

                // Is there something to post
                if (!string.IsNullOrEmpty(content))
                {

                    int postLimitLength = _postLengthLimit;

                    // First shorten the URL, if it exists.
                    string shortenedURL = "";
                    if (trackbackUri is not null)
                    {
                        shortenedURL = ShortenURL(trackbackUri.ToString());
                        // If shortening has failed check the original uri
                        if (string.IsNullOrEmpty(shortenedURL))
                            shortenedURL = trackbackUri.ToString();
                        if (!string.IsNullOrEmpty(shortenedURL))
                            shortenedURL = " " + shortenedURL;
                    }

                    postLimitLength -= shortenedURL.Length;

                    // Truncate the content
                    content = Tools.Text.TruncateString(content, postLimitLength);

                    Update(content + shortenedURL);

                    this.myWeb.moDbHelper.logActivity(Cms.dbHelper.ActivityType.IntegrationTwitterPost, this.DirectoryId, 0, contentId, shortenedURL);

                }
            }

            catch (Exception ex)
            {
                OnError?.Invoke(this, new Tools.Errors.ErrorEventArgs(this._moduleName, "Post()", ex, ""));
            }
        }

        public class User
        {
            public string id
            {
                get
                {
                    return m_id;
                }
                set
                {
                    m_id = value;
                }
            }
            private string m_id;
            public string first_name
            {
                get
                {
                    return m_first_name;
                }
                set
                {
                    m_first_name = value;
                }
            }
            private string m_first_name;
            public string last_name
            {
                get
                {
                    return m_last_name;
                }
                set
                {
                    m_last_name = value;
                }
            }
            private string m_last_name;
            public string email
            {
                get
                {
                    return m_email;
                }
                set
                {
                    m_email = value;
                }
            }
            private string m_email;
            public string link
            {
                get
                {
                    return m_link;
                }
                set
                {
                    m_link = value;
                }
            }
            private string m_link;
            public string username
            {
                get
                {
                    return m_username;
                }
                set
                {
                    m_username = value;
                }
            }
            private string m_username;
            public string gender
            {
                get
                {
                    return m_gender;
                }
                set
                {
                    m_gender = value;
                }
            }
            private string m_gender;
            public string locale
            {
                get
                {
                    return m_locale;
                }
                set
                {
                    m_locale = value;
                }
            }
            private string m_locale;
        }

    }



}